Optimism Wallet Summary - WIP

henrystats opened the pull request:
# Thank you for contributing to Spellbook!
Please refer to the top of the `readme` in the root of Spellbook to learn how to contribute to Spellbook on DuneSQL.

dune-eng commented:
Workflow run id 5912057508 approved.

dune-eng commented:
Workflow run id 5912057553 approved.

dune-eng commented:
Workflow run id 5913926397 approved.

dune-eng commented:
Workflow run id 5913926507 approved.

dune-eng commented:
Workflow run id 5920002394 approved.

dune-eng commented:
Workflow run id 5920002426 approved.

dune-eng commented:
Workflow run id 5920042300 approved.

dune-eng commented:
Workflow run id 5920042310 approved.

dune-eng commented:
Workflow run id 5920099225 approved.

dune-eng commented:
Workflow run id 5920099303 approved.

dune-eng commented:
Workflow run id 5920169361 approved.

dune-eng commented:
Workflow run id 5920169412 approved.

dune-eng commented:
Workflow run id 5920254171 approved.

dune-eng commented:
Workflow run id 5920254185 approved.

dune-eng commented:
Workflow run id 5920304477 approved.

dune-eng commented:
Workflow run id 5920304460 approved.

henrystats commented:
depends on pr 4130

chuxinh commented on models/addresses_summary/optimism/addresses_summary_optimism_mainet.sql:
A few other things could be helpful to add:
- First contract interacted, or even like a path of contracts interacted for the first few transactions? You can you contract_mapping table for that 
- First function call 
- address recency 
- amount of op delegation (first delegated at block_number, block_time, amount etc). I previously created a table in Dune trying to do the address summary. Feel free to borrow some ideas from there: https://dune.com/queries/2794862

henrystats replied:
okee.. thank you... I'll resolve your comments and also add the other things later this evening.

chuxinh commented on models/addresses_summary/optimism/addresses_summary_optimism_schema.yml:
```diff
@@ -0,0 +1,35 @@
+version: 2
+
+models:
+  - name: addresses_summary_optimism_mainet
```
```suggestion
  - name: addresses_summary_op_mainnet
```

chuxinh commented on models/addresses_summary/optimism/addresses_summary_optimism_schema.yml:
```diff
@@ -0,0 +1,35 @@
+      contributors: Henrystats
+    config:
+      tags: ['table', 'wallet summary', 'mainet', 'addresses_summary', 'optimism']
+    description: "Table showing summary of an address on Optimism mainet"
```
```suggestion
    description: "Table showing summary of an address on OP mainnet"
```

henrystats commented on models/addresses_summary/optimism/addresses_summary_optimism_mainet.sql:
```diff
@@ -0,0 +1,106 @@
+    END as usage_frequency,
+    COUNT(ot.hash) as number_of_transactions,
+    COUNT(DISTINCT(cm.contract_project)) as unique_dapps 
+FROM 
```
yep the latter.. the table only updates data for an address that's included in the weekly active.

dune-eng commented:
Workflow run id 5946158581 approved.

dune-eng commented:
Workflow run id 5946158465 approved.

dune-eng commented:
Workflow run id 5946165059 approved.

dune-eng commented:
Workflow run id 5946165119 approved.

dune-eng commented:
Workflow run id 5946302484 approved.

dune-eng commented:
Workflow run id 5946302403 approved.

dune-eng commented:
Workflow run id 5946306927 approved.

dune-eng commented:
Workflow run id 5946307010 approved.

dune-eng commented:
Workflow run id 5946469096 approved.

dune-eng commented:
Workflow run id 5946469115 approved.

dune-eng commented:
Workflow run id 5946688677 approved.

dune-eng commented:
Workflow run id 5946688732 approved.

dune-eng commented:
Workflow run id 5946733253 approved.

dune-eng commented:
Workflow run id 5946733307 approved.

dune-eng commented:
Workflow run id 5946927872 approved.

dune-eng commented:
Workflow run id 5946927916 approved.

dune-eng commented:
Workflow run id 5946963276 approved.

dune-eng commented:
Workflow run id 5946963330 approved.

dune-eng commented:
Workflow run id 5947003087 approved.

dune-eng commented:
Workflow run id 5947003134 approved.

dune-eng commented:
Workflow run id 5947106856 approved.

dune-eng commented:
Workflow run id 5947106898 approved.

dune-eng commented:
Workflow run id 5947190219 approved.

henrystats commented:
@chuxinh I've made the changes you suggested and added the new columns as well... two things I couldn't add:

1. path of first few transactions, I add the first contract project & first contract address, but I didn't add the path because the only way I could of getting that data (the array) is to use row_number over the address and then filter the rank to get the first 3 contracts..

that would take a lot of compute resource, I was thinking it'd be better to set up a separate spell similar to first activity to get the path, and think of a way to design that incrementally... row_number on the entire optimism transactions table sound like a lot of compute resource... or what do you think?

2. delegate data (I believe there's another dework bounty up to add delegate data to spellbook, soo it's best to wait for that to get set up and I can pull data from there.. similar to the bridge flows data which I have a pr set up for... 


Also, I removed two signature id's because they were causing duplicates as they have multiple functions...

`AND id NOT IN (0x09779838, 0x00000000)`

chuxinh commented on models/addresses_events/optimism/addresses_events_optimism_first_activity.sql:
```diff
@@ -0,0 +1,70 @@
+WHERE 1 = 1
+{% endif %}
+{% if is_incremental() %}
+AND et.block_time >= date_trunc('day', now() - interval '7' day)
```
move the block_time condition inside the subquery instead?

chuxinh commented on models/addresses_events/optimism/addresses_events_optimism_first_activity.sql:
```diff
@@ -0,0 +1,70 @@
+SELECT 'optimism' AS blockchain
+, et."from" AS address
+, MIN_BY(et."to", et.block_number) AS first_activity_to
+, MIN(et.block_time) AS block_time
+, MIN(et.block_number) AS block_number
+, MIN_BY(et.hash, et.block_number) AS tx_hash
```
should we also add `first_` prefix to be consistent?

chuxinh commented on models/addresses_events/optimism/addresses_events_optimism_schema.yml:
```diff
@@ -36,3 +36,46 @@ models:
+      contributors: Henrystats
+    config:
+      tags: ['table', 'activity', 'addresses_events', 'optimism']
+    description: "Table showing first activity of each Optimism address in ETH"
```
```suggestion
    description: "Table showing first activity of each OP Mainnet address"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: blockchain
+        description: "Blockchain"
+      - name: address
+        description: "Concerned Optimism address"
```
```suggestion
        description: "Concerned OP Mainnet address"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: last_active_time
+        description: "The block time on which the address made it's last OP mainet transaction" 
+      - name: address_age
+        description: "Time period the address has been active for labelled " 
```
```suggestion
        description: "Time period the address has been active for labelled" 
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: address_age_in_days
+        description: "Time period in days the address has been active for " 
+      - name: recency_age_in_days
+        description: "Time period in days since the address was last active " 
```
```suggestion
        description: "Time period in days since the address was last active" 
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      contributors: Henrystats
+    config:
+      tags: ['table', 'wallet summary', 'mainet', 'addresses_summary', 'optimism']
+    description: "Table showing summary of an address on OP mainet"
```
```suggestion
    description: "Table showing summary of an address on OP Mainnet"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: first_active_time
+        description: "The block time on which the address made it's first OP mainet transaction"
+      - name: first_transaction_hash
+        description: "The transaction hash on which the address made it's first OP mainet transaction"
```
```suggestion
        description: "The transaction hash on which the address made it's first OP Mainnet transaction"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: first_function
+        description: "first function called by the address "
+      - name: last_active_time
+        description: "The block time on which the address made it's last OP mainet transaction" 
```
```suggestion
        description: "The block time on which the address made it's last OP Mainnet transaction" 
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: address_age
+        description: "Time period since the address was last active for labelled " 
+      - name: address_age_in_days
+        description: "Time period in days the address has been active for " 
```
```suggestion
        description: "Time period in days the address has been active for" 
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: recency_age_in_days
+        description: "Time period in days since the address was last active " 
+      - name: number_of_transactions
+        description: "Number of mainet transactions made by the address"
```
```suggestion
        description: "Number of OP Mainnet transactions made by the address"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: usage_frequency
+        description: "How often address is active"
+      - name: gas_spent
+        description: "total gas spent by adddress on OP mainet"
```
```suggestion
        description: "Total gas spent by an address on OP Mainnet"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: gas_spent
+        description: "total gas spent by adddress on OP mainet"
+      - name: unique_dapps
+        description: "unique dapps address interacted with OP mainet"
```
```suggestion
        description: "Unique dapps addresses interacted with OP Mainnet"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: unique_dapps
+        description: "unique dapps address interacted with OP mainet"
+      - name: first_to_project
+        description: "project address sent first transaction to "
```
```suggestion
        description: "Project address sent first transaction to"
```

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: first_to_project
+        description: "project address sent first transaction to "
+      - name: first_to_address
+        description: "contract address, address sent first transaction to "
```
```suggestion
        description: "Contract address that an address sent first transaction to"
```

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,216 @@
+            to, 
+            hash,
+            block_time, 
+            (l1_fee + (gas_used * gas_price))/1e18 as gas_spent,
```
```suggestion
            case when gas_price = cast(0 as uint256) then 0
            else cast(gas_used as double) * cast(gas_price as double)/1e18 + cast(l1_fee as double) /1e18
            end as gas_spent,
```

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,216 @@
+LEFT JOIN 
+{{ ref('contracts_optimism_contract_mapping') }} cm 
+    ON ot."to" = cm.contract_address 
+LEFT JOIN (
+    SELECT 
+        DISTINCT id, 
+        split_part(signature,'(',1) as function 
+    FROM 
+    {{ ref('signatures') }} 
+    where type = 'function_call'
+    AND id NOT IN (0x09779838, 0x00000000) -- for some weird reason these have duplicates functions
+) sig 
+    ON sig.id = fa.first_function
```
maybe this part can be done in `addresses_events_optimism_first_activity` instead of here?

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,216 @@
+            to, 
+            hash,
+            block_time, 
+            (l1_fee + (gas_used * gas_price))/1e18 as gas_spent,
```
modify as suggested above

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,216 @@
+LEFT JOIN 
+{{ ref('contracts_optimism_contract_mapping') }} cm 
+    ON ot."to" = cm.contract_address 
+LEFT JOIN (
```
modify as suggested above

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,216 @@
+    SUM(gas_spent) as gas_spent,
+    MAX(ot.block_time) as last_active_time, 
+    CASE
+           WHEN date_diff('day', min(ot.block_time), max(ot.block_time)) > 1825 THEN '5 years old User'
```
can use the first_active_time here?

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,216 @@
+           WHEN date_diff('day', max(ot.block_time), CAST(NOW() as timestamp)) > 7 THEN 'Last active more than 1 week Ago'
+           ELSE 'Active within the past week'
+    END as recency_age, 
+    date_diff('day', min(ot.block_time), max(ot.block_time)) as address_age_in_days,
```
same here that can use the first_active_time

dune-eng commented:
Workflow run id 5953276863 approved.

dune-eng commented:
Workflow run id 5953277087 approved.

dune-eng commented:
Workflow run id 5953450960 approved.

dune-eng commented:
Workflow run id 5953560825 approved.

henrystats commented:
@chuxinh I've made the changes as requested... for the first_contract_project did you mean I should add that in first_activity as well? 

Will do that in a bit if that's the case, originally left it out since I assumed you suggested that so we'd not need to do a left join with the contract mapping table but we need to because of the unique dapps column.. 

ci is failing now cause a couple of heavy pr's recently got merged.

henrystats commented on models/addresses_summary/optimism/addresses_summary_optimism_mainet.sql:
```diff
@@ -0,0 +1,106 @@
+            WHEN (COUNT(ot.hash)/(date_diff('day', min(ot.block_time), max(ot.block_time)) + 1)) >= 1 THEN 'Daily User'
+            WHEN (COUNT(ot.hash)/(date_diff('day', min(ot.block_time), max(ot.block_time)) + 1)) >= 0.142857142857 THEN 'Weekly User'
+            WHEN (COUNT(ot.hash)/(date_diff('day', min(ot.block_time), max(ot.block_time)) + 1)) >= 0.0333333333333 THEN 'Monthly User'
+            WHEN (COUNT(ot.hash)/(date_diff('day', min(ot.block_time), max(ot.block_time)) + 1)) >= 0.0027397260274 THEN 'Yearly User'
```
number of transactions / number of days active (age)

soo for a daily user they need to  make at least one transaction per day 

for a weekly user they need to make at least one transaction per week which 1/7 = 0.142857.. same for monthly and yearly as well.

henrystats replied:
I added + 1 because date_diff returns zero (causes an error since we're dividing by 0) if an address as only been active on one day... this is causing the label to be inaccurate when an address is only active one day so will fix that shortly.

dune-eng commented:
Workflow run id 5958481062 approved.

dune-eng commented:
Workflow run id 5958481058 approved.

chuxinh commented:
> 

Oh I was referring to the function signature since you only use this for the first tx

henrystats commented:
ohh oke i had updated that @chuxinh

jeff-dude commented:
updated to `wip` -- feel free to edit title & tag when ready for final review

henrystats commented:
thank you ser!

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+        description: "first function called by the address"
+      - name: last_active_time
+        description: "The block time on which the address made it's last OP Mainnet transaction" 
+      - name: address_age
+        description: "Time period the address has been active for labelled" 
+      - name: address_age
+        description: "Time period since the address was last active for labelled" 
```
duplicate

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+        description: "Time period since the address was last active for labelled" 
+      - name: address_age_in_days
+        description: "Time period in days the address has been active for" 
+      - name: recency_age_in_days
```
```suggestion
      - name: recency_in_days
```

chuxinh replied:
maybe change to the name above. also remember to update that in sql

chuxinh commented on models/addresses_summary/addresses_summary_schema.yml:
```diff
@@ -0,0 +1,47 @@
+      - name: usage_frequency
+        description: "How often address is active"
+      - name: gas_spent
+        description: "Total gas spent by an address on OP Mainnet"
```
```suggestion
        description: "Total gas spent in ETH by an address on OP Mainnet"
```

chuxinh commented on models/addresses_events/optimism/addresses_events_optimism_schema.yml:
```diff
@@ -36,3 +36,44 @@ models:
+        tests:
+          - unique
+      - name: first_activity_to
+        description: "Which Optimism address received the first activity of the address with ETH"
```
First transaction activity to address

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,194 @@
+            case when gas_price = cast(0 as uint256) then 0
+            else cast(gas_used as double) * cast(gas_price as double)/1e18 + cast(l1_fee as double) /1e18
+            end as gas_spent,
+            data, 
```
remove data column if it's not used

chuxinh commented on models/addresses_summary/addresses_summary_op_mainnet.sql:
```diff
@@ -0,0 +1,194 @@
+    END as usage_frequency,
+    COUNT(ot.hash) as number_of_transactions,
+    COUNT(DISTINCT(cm.contract_project)) as unique_dapps,
+    MIN_BY(cm.contract_project, ot.block_number) as first_to_project, 
+    MIN_BY(ot.to, ot.block_number) as first_to_address
```
maybe can also at the last  to address and project?

chuxinh reviewed (changes requested):
Logic looks good to me! Just a few more comments and thanks so much for this!! @henrystats

dune-eng commented:
Workflow run id 5970594069 approved.

dune-eng commented:
Workflow run id 5970594186 approved.

dune-eng commented:
Workflow run id 5970595022 approved.

dune-eng commented:
Workflow run id 5970595097 approved.

henrystats commented:
made the changes @chuxinh
